File: lib/algorithms/binary_search/binary_search.js
function binarySearch(array, val) {
  const length = array.length;

  if (length <= 0) return;
  if (length == 1 && array[0] === val) return array[0];
  if (length == 1 && array[0] !== val) return;

  let low = 0;
  let high = length;
  let mid;

  while (low <= high) {
    mid = Math.floor((low + high) / 2);

    if (array[mid] === val) return mid;

    if (array[mid] < val) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
}

console.log(binarySearch([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 3));

File: lib/algorithms/binary_search/binary_search.rb
def binary_search(array, val)
  return if array.empty? || array.nil?
  return if array.one? && array.first != val
  return array.first if array.one?

  low = 0
  high = array.length - 1

  while low <= high
    mid = (low + high) / 2

    if array[mid] == val
      return mid
    elsif array[mid] > val
      high = mid - 1
    else
      low = mid + 1
    end
  end

  nil
end
